eth_p2p/blockchain_sync.nim

```diff
@@ -72,25 +72,31 @@ proc availableWorkItem(ctx: SyncContext): int =
   ctx.workQueue[result] = WantedBlocks(startIndex: nextRequestedBlock, numBlocks: numBlocks.uint, state: Initial)
 
 proc persistWorkItem(ctx: SyncContext, wi: var WantedBlocks) =
-  ctx.chain.persistBlocks(wi.headers, wi.bodies)
+  case ctx.chain.persistBlocks(wi.headers, wi.bodies)
+  of ValidationResult.OK:
+    ctx.finalizedBlock = wi.endIndex
+    wi.state = Persisted
+  of ValidationResult.RecoverableValidationError:
```

Comment:
I think the recoverable errors should be as descriptive as possible, but it should be also easy for the user to write code that handles just the two conditions "error happened" and "operation completed without errors". To achieve this with error codes, it's easy as:

``` nim
if operation() != Success:
  ...
```

With exceptions, you can define a common base type for all the errors of your library (e.g. see `RlpError` from `nim-rlp` from example) and then have more specific recoverable error types derived from it.